src/features/device-state/routes/state-patch-v3.ts

```diff
@@ -0,0 +1,384 @@
+			const resinApiTx = api.resin.clone({
+				passthrough: { req, custom, tx },
+			});
+
+			const waitPromises: Array<PromiseLike<any>> = [];
+
+			for (const uuid of uuids) {
+				const state = body[uuid];
+
+				const { apps } = state;
+
+				let deviceBody:
+					| undefined
+					| (Pick<
+							StatePatchV3Body[string],
+							typeof v3ValidPatchFields[number]
+					  > & {
+							device_name?: string;
+							is_running__release?: number | null;
+					  }) = _.pick(state, v3ValidPatchFields);
+				let metricsBody:
+					| undefined
+					| Pick<StatePatchV3Body[string], typeof metricsPatchFields[number]> =
+					_.pick(state, metricsPatchFields);
+				if (
+					Object.keys(metricsBody).length > 0 &&
+					(await shouldUpdateMetrics(uuid))
+				) {
+					// If we should force a metrics update then merge the two together and clear `metricsBody` so
+					// that we don't try to merge it again later
+					deviceBody = { ...deviceBody, ...metricsBody };
+					metricsBody = {};
+				}
+
+				if (state.name != null) {
+					deviceBody.device_name = state.name;
+				}
+
+				const device = devices.find((d) => d.uuid === uuid);
+				if (device == null) {
+					throw new UnauthorizedError();
+				}
+				const userAppUuid = device.belongs_to__application[0].uuid;
+				if (apps != null) {
+					const release = appReleases[userAppUuid].find(
+						(r) => r.commit === apps[userAppUuid].release_uuid,
+					);
+					if (release) {
+						deviceBody.is_running__release = release.id;
+					}
+				}
+
+				if (!_.isEmpty(deviceBody)) {
+					// If we're updating anyway then ensure the metrics data is included
+					deviceBody = { ...deviceBody, ...metricsBody };
+					waitPromises.push(
+						resinApiTx.patch({
+							resource: 'device',
+							id: device.id,
+							options: {
+								$filter: { $not: deviceBody },
+							},
+							body: deviceBody,
+						}),
+					);
+				}
+
+				if (apps != null) {
+					const imgInstalls: Array<{
+						imageId: number;
+						releaseId: number;
+						status: string;
+						downloadProgress?: number;
+					}> = [];
+					for (const [
+						appUuid,
+						{
+							// release_uuid: isRunningReleaseUuid,
+							releases = {},
+						},
+					] of Object.entries(apps)) {
+						// // TODO: This gets the release we are running for the given app but currently we handle the user app out of band above, and ignore supervisor/os
+						// const release = releases[appUuid].find(
+						// 	(r) => r.commit === isRunningReleaseUuid,
+						// );
+						// if (release == null) {
+						// 	throw new InternalRequestError();
+						// }
+						for (const [releaseUuid, { services = {} }] of Object.entries(
+							releases,
+						)) {
+							const release = appReleases[appUuid].find(
+								(r) => r.commit === releaseUuid,
+							);
+							if (release == null) {
+								throw new InternalRequestError();
+							}
+							for (const service of Object.values(services)) {
+								const image = images.find(
```

Comment:
I guess we could have made the images be a dictionary, but of course no big deal...